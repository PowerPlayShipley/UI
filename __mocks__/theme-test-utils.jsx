// TODO: Fix this??

import * as React from 'react'
import userEvent from '@testing-library/user-event'
import { render as rtlRender } from '@testing-library/react'

import { ThemeProvider } from "styled-components";

import theme from '../src/themes'
import GlobalStyle from "../src/components/GlobalStyle";

function renderWithTheme(ui, { ...options } = {}) {
  const Wrapper = ({ children }) => (
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
      <div id='modal-root' />
    </ThemeProvider>
  )

  return rtlRender(ui, { wrapper: Wrapper, ...options })
}

// THE BELLOW COME FROM `https://github.com/styled-components/styled-components/blob/0aa3170c255a49cd41c3fbeb2b8051b5d132f229/src/test/utils.js`

const stripComments = (str) => str.replace(/\/\*.*?\*\/\n?/g, '');

const stripWhitespace = (str) => str
  .trim()
  .replace(/([;\{\}])/g, '$1  ')
  .replace(/\s+/g, ' ');

const getCSS = (scope) =>
  Array.from(scope.querySelectorAll('style'))
    .map(tag => tag.innerHTML)
    .join('\n')
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

function expectCSSMatches ( _expectation, opts = { ignoreWhitespace: true }) {
  // NOTE: This should normalise both CSS strings to make irrelevant mismatches less likely
  const expectation = _expectation
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

  const css = getCSS(document);

  if (opts.ignoreWhitespace) {
    const stripped = stripWhitespace(stripComments(css));
    expect(stripped).toEqual(stripWhitespace(expectation));
    return stripped;
  } else {
    expect(css).toEqual(expectation);
    return css;
  }
};

export * from '@testing-library/react';

export { renderWithTheme, expectCSSMatches };
