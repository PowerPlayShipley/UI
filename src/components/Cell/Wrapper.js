import styled from "styled-components";

const Wrapper = styled.span`
  display: block;
`

export default Wrapper
