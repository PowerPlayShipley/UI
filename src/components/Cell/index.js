import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'

import Wrapper from "./Wrapper";

// Going to keep these so we can add them to the span bellow just for prosperity
const Cell = ({ value: initialValue, row: { index }, column: { id }}) => {
  // We need to keep and update the state of the cell normally
  const [value, setValue] = useState(initialValue)

  // If the initialValue is changed external, sync it up with our state
  useEffect(() => {
    setValue(initialValue)
  }, [initialValue])

  return <Wrapper data-testid="cell-value" data-row={index} data-column={id}>{value}</Wrapper>
}

Cell.propTypes = {
  value: PropTypes.string.isRequired,
  row: PropTypes.shape({
    index: PropTypes.number
  }).isRequired,
  column: PropTypes.shape({
    id: PropTypes.number
  }).isRequired
}

export default Cell
