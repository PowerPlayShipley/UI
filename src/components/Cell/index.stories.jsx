import Cell from "./index";

export default {
  title: 'Cell/Cell',
  component: Cell
}

const Template = ({row, column, ...rest}) => {
  return (<Cell row={{ index: row }} column={{ id: column }} {...rest} />);
}

export const Default = Template.bind({});
Default.args = {
  row: 1,
  column: 1,
  value: ['A', 'B', 'C'].join(' ')
}
