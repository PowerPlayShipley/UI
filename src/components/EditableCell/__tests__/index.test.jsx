/**
 * @jest-environment jsdom
 */

import React from 'react'
import { renderWithTheme, fireEvent } from '../../../../__mocks__/theme-test-utils'

import Cell from "../index";

describe('Editable Cell', () => {
  it('should render cell', () => {
    const { container } = renderWithTheme(<Cell column={{ id: 1 }} row={{ index: 1 }} value={['Hello']} />)
    expect(container).toHaveTextContent('Hello')

    const svg = container.querySelector('svg')
    expect(svg).toHaveClass('fa-donate')
  });

  it('should render disabled button', function () {
    const { container } = renderWithTheme(<Cell column={{ id: 1 }} row={{ index: 1 }} value={['Hello']} disabled={true} />)
    expect(container).toHaveTextContent('Hello')

    const button = container.querySelector('button')
    expect(button).toBeDisabled()
  });

  it('should handle toolbar click', function () {
    const handleClick = jest.fn()

    const { container } = renderWithTheme(<Cell column={{ id: 1 }} row={{ index: 1 }} value={['Hello']} onToolbarClick={handleClick} />)

    fireEvent.click(container.querySelector('svg'))
    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  it('should handle item click', function () {
    const handleClick = jest.fn()

    const { getByText } = renderWithTheme(<Cell column={{ id: 1 }} row={{ index: 1 }} value={['Hello']} onItemClick={handleClick} />)

    fireEvent.click(getByText('Hello').parentNode)
    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  it('should create valid snapshot', function () {
    const { container } = renderWithTheme(<Cell column={{ id: 1 }} row={{ index: 1 }} value={['Hello']} />)
    expect(container.firstChild).toMatchSnapshot()
  });
})
