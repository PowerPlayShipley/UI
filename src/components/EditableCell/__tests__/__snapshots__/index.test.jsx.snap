// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editable Cell should create valid snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 0 0;
}

.c0 div:first-child {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 div[disabled],
.c1 div.disabled,
.c1 div[aria-disabled=true] {
  pointer-events: none;
  color: #8c959f;
  opacity: 65%;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 24px;
  min-width: 24px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 3px;
  padding: 4px;
  color: #222f3e;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c3:hover {
  background-color: #dee0e2;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  max-height: 200px;
  overflow-x: hidden;
  overflow-y: auto;
  padding: 0;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        aria-disabled="false"
        class="c3"
        data-item="[object Object]"
      >
        <div
          class="c4"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-donate fa-w-16 "
            data-icon="donate"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 416c114.9 0 208-93.1 208-208S370.9 0 256 0 48 93.1 48 208s93.1 208 208 208zM233.8 97.4V80.6c0-9.2 7.4-16.6 16.6-16.6h11.1c9.2 0 16.6 7.4 16.6 16.6v17c15.5.8 30.5 6.1 43 15.4 5.6 4.1 6.2 12.3 1.2 17.1L306 145.6c-3.8 3.7-9.5 3.8-14 1-5.4-3.4-11.4-5.1-17.8-5.1h-38.9c-9 0-16.3 8.2-16.3 18.3 0 8.2 5 15.5 12.1 17.6l62.3 18.7c25.7 7.7 43.7 32.4 43.7 60.1 0 34-26.4 61.5-59.1 62.4v16.8c0 9.2-7.4 16.6-16.6 16.6h-11.1c-9.2 0-16.6-7.4-16.6-16.6v-17c-15.5-.8-30.5-6.1-43-15.4-5.6-4.1-6.2-12.3-1.2-17.1l16.3-15.5c3.8-3.7 9.5-3.8 14-1 5.4 3.4 11.4 5.1 17.8 5.1h38.9c9 0 16.3-8.2 16.3-18.3 0-8.2-5-15.5-12.1-17.6l-62.3-18.7c-25.7-7.7-43.7-32.4-43.7-60.1.1-34 26.4-61.5 59.1-62.4zM480 352h-32.5c-19.6 26-44.6 47.7-73 64h63.8c5.3 0 9.6 3.6 9.6 8v16c0 4.4-4.3 8-9.6 8H73.6c-5.3 0-9.6-3.6-9.6-8v-16c0-4.4 4.3-8 9.6-8h63.8c-28.4-16.3-53.3-38-73-64H32c-17.7 0-32 14.3-32 32v96c0 17.7 14.3 32 32 32h448c17.7 0 32-14.3 32-32v-96c0-17.7-14.3-32-32-32z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        aria-disabled="false"
        class="c3"
        data-item="Hello"
      >
        <div
          class="c4"
        >
          Hello
        </div>
      </div>
    </div>
  </div>
</div>
`;
